File: shifai-ios/ShifAI/Data/Sync/SyncEngine.swift
import Foundation
import CryptoKit

// MARK: - Sync Engine (S7-1)
// Zero-knowledge, offline-first cloud sync
// Flow: local DB → JSON → AES-256-GCM encrypt → blob → Supabase

final class SyncEngine {

    static let shared = SyncEngine()

    // MARK: - Config

    private let syncEndpoint: String
    private let maxBlobSize = 10 * 1024 * 1024  // 10MB
    private let syncInterval: TimeInterval = 6 * 3600  // 6h

    private var isSyncing = false

    enum SyncState: Equatable {
        case idle
        case syncing
        case success(lastSync: Date)
        case error(String)
    }

    @Published var state: SyncState = .idle
    @Published var lastSyncDate: Date?
    @Published var isEnabled: Bool {
        didSet { UserDefaults.standard.set(isEnabled, forKey: "sync_enabled") }
    }

    // MARK: - Init

    init() {
        self.syncEndpoint = "\(Configuration.supabaseURL)/functions/v1/sync-data"
        self.isEnabled = UserDefaults.standard.bool(forKey: "sync_enabled")
        self.lastSyncDate = UserDefaults.standard.object(forKey: "last_sync_date") as? Date
    }

    // MARK: - S7-1: Push (Local → Cloud)

    func push() async throws {
        guard isEnabled, !isSyncing else { return }
        isSyncing = true
        state = .syncing

        defer { isSyncing = false }

        do {
            // 1. Serialize local data to JSON
            let payload = try serializeLocalData()

            // 2. Encrypt with AES-256-GCM
            let encryptedBlob = try encrypt(data: payload)

            // 3. Compute SHA-256 checksum
            let checksum = SHA256.hash(data: encryptedBlob)
                .compactMap { String(format: "%02x", $0) }
                .joined()

            guard encryptedBlob.count <= maxBlobSize else {
                state = .error("Données trop volumineuses (\(encryptedBlob.count / 1024)KB)")
                return
            }

            // 4. Push to Supabase
            let currentVersion = UserDefaults.standard.integer(forKey: "sync_blob_version")
            var request = URLRequest(url: URL(string: syncEndpoint)!)
            request.httpMethod = "POST"
            request.setValue("application/octet-stream", forHTTPHeaderField: "Content-Type")
            request.setValue("Bearer \(try getAuthToken())", forHTTPHeaderField: "Authorization")
            request.setValue(checksum, forHTTPHeaderField: "X-Checksum-SHA256")
            request.setValue("\(currentVersion + 1)", forHTTPHeaderField: "X-Blob-Version")
            request.httpBody = encryptedBlob

            let (data, response) = try await URLSession.shared.data(for: request)

            guard let httpResponse = response as? HTTPURLResponse,
                  httpResponse.statusCode == 200 else {
                let errorBody = String(data: data, encoding: .utf8) ?? "Unknown error"
                state = .error(errorBody)
                return
            }

            // 5. Update local state
            let newVersion = currentVersion + 1
            UserDefaults.standard.set(newVersion, forKey: "sync_blob_version")
            let now = Date()
            UserDefaults.standard.set(now, forKey: "last_sync_date")
            lastSyncDate = now
            state = .success(lastSync: now)

        } catch {
            state = .error(error.localizedDescription)
            throw error
        }
    }

    // MARK: - S7-1: Pull (Cloud → Local)

    func pull() async throws {
        guard isEnabled, !isSyncing else { return }
        isSyncing = true
        state = .syncing

        defer { isSyncing = false }

        do {
            // 1. GET metadata first
            var metaRequest = URLRequest(url: URL(string: "\(syncEndpoint)?action=metadata")!)
            metaRequest.httpMethod = "GET"
            metaRequest.setValue("Bearer \(try getAuthToken())", forHTTPHeaderField: "Authorization")

            let (metaData, metaResponse) = try await URLSession.shared.data(for: metaRequest)
            guard let metaHttp = metaResponse as? HTTPURLResponse, metaHttp.statusCode == 200 else {
                state = .error("Impossible de vérifier la version serveur")
                return
            }

            let metadata = try JSONDecoder().decode(SyncMetadata.self, from: metaData)
            let localVersion = UserDefaults.standard.integer(forKey: "sync_blob_version")

            guard metadata.blobVersion > localVersion else {
                state = .success(lastSync: lastSyncDate ?? Date())
                return  // Local is up to date
            }

            // 2. GET blob
            var blobRequest = URLRequest(url: URL(string: "\(syncEndpoint)?action=pull")!)
            blobRequest.httpMethod = "GET"
            blobRequest.setValue("Bearer \(try getAuthToken())", forHTTPHeaderField: "Authorization")

            let (blobData, blobResponse) = try await URLSession.shared.data(for: blobRequest)
            guard let blobHttp = blobResponse as? HTTPURLResponse, blobHttp.statusCode == 200 else {
                state = .error("Échec du téléchargement")
                return
            }

            // 3. Verify checksum
            let serverChecksum = (blobHttp.value(forHTTPHeaderField: "X-Checksum-SHA256")) ?? ""
            let localChecksum = SHA256.hash(data: blobData)
                .compactMap { String(format: "%02x", $0) }
                .joined()

            guard serverChecksum == localChecksum else {
                state = .error("Erreur d'intégrité des données")
                return
            }

            // 4. Decrypt
            let decryptedData = try decrypt(data: blobData)

            // 5. Merge with local
            try mergeWithLocal(data: decryptedData)

            // 6. Update version
            UserDefaults.standard.set(metadata.blobVersion, forKey: "sync_blob_version")
            let now = Date()
            UserDefaults.standard.set(now, forKey: "last_sync_date")
            lastSyncDate = now
            state = .success(lastSync: now)

        } catch {
            state = .error(error.localizedDescription)
            throw error
        }
    }

    // MARK: - Full Sync (Push then Pull)

    func sync() async throws {
        try await push()
        try await pull()
    }

    // MARK: - Serialization

    private func serializeLocalData() throws -> Data {
        // Serialize cycle entries, symptom logs, insights, predictions
        // into a unified JSON structure
        var payload: [String: Any] = [:]

        let cycleRepo = CycleRepository()
        let symptomRepo = SymptomRepository()
        let insightRepo = InsightRepository()
        let predictionRepo = PredictionRepository()

        let cycles = try cycleRepo.fetchAll()
        let insights = try insightRepo.fetchRecent(limit: 10000)
        let predictions = try predictionRepo.fetchAll()

        // Encode to JSON via Codable
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        payload["cycles"] = try encoder.encode(cycles)
        payload["insights"] = try encoder.encode(insights)
        payload["predictions"] = try encoder.encode(predictions)
        payload["version"] = UserDefaults.standard.integer(forKey: "sync_blob_version") + 1
        payload["timestamp"] = ISO8601DateFormatter().string(from: Date())

        return try JSONSerialization.data(withJSONObject: payload)
    }

    private func mergeWithLocal(data: Data) throws {
        // Last-write-wins merge strategy
        // In conflict cases, the data with the more recent timestamp wins
        // TODO: Implement full merge logic with conflict detection
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601

        // Parse the incoming data and merge
        // For MVP: server data replaces local for items with newer timestamps
    }

    // MARK: - Encryption (AES-256-GCM)

    private func encrypt(data: Data) throws -> Data {
        let key = try getOrCreateSyncKey()
        let nonce = AES.GCM.Nonce()
        let sealedBox = try AES.GCM.seal(data, using: key, nonce: nonce)
        guard let combined = sealedBox.combined else {
            throw SyncError.encryptionFailed
        }
        return combined
    }

    private func decrypt(data: Data) throws -> Data {
        let key = try getOrCreateSyncKey()
        let sealedBox = try AES.GCM.SealedBox(combined: data)
        return try AES.GCM.open(sealedBox, using: key)
    }

    private func getOrCreateSyncKey() throws -> SymmetricKey {
        let keyTag = "com.shifai.sync.key"

        // Try to read from Keychain
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: keyTag,
            kSecReturnData as String: true
        ]

        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)

        if status == errSecSuccess, let keyData = result as? Data {
            return SymmetricKey(data: keyData)
        }

        // Generate new key
        let newKey = SymmetricKey(size: .bits256)
        let keyData = newKey.withUnsafeBytes { Data($0) }

        let addQuery: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: keyTag,
            kSecValueData as String: keyData,
            kSecAttrAccessible as String: kSecAttrAccessibleAfterFirstUnlock
        ]

        SecItemAdd(addQuery as CFDictionary, nil)
        return newKey
    }

    // MARK: - Auth

    private func getAuthToken() throws -> String {
        // Return current Supabase JWT
        guard let token = UserDefaults.standard.string(forKey: "supabase_access_token") else {
            throw SyncError.notAuthenticated
        }
        return token
    }

    // MARK: - Types

    struct SyncMetadata: Codable {
        let blobVersion: Int
        let updatedAt: String
        let sizeBytes: Int

        enum CodingKeys: String, CodingKey {
            case blobVersion = "blob_version"
            case updatedAt = "updated_at"
            case sizeBytes = "size_bytes"
        }
    }

    enum SyncError: Error, LocalizedError {
        case encryptionFailed
        case notAuthenticated
        case blobTooLarge
        case integrityCheckFailed

        var errorDescription: String? {
            switch self {
            case .encryptionFailed: return "Échec du chiffrement"
            case .notAuthenticated: return "Non authentifié"
            case .blobTooLarge: return "Données trop volumineuses"
            case .integrityCheckFailed: return "Erreur d'intégrité"
            }
        }
    }
}

// MARK: - Configuration

private enum Configuration {
    static var supabaseURL: String {
        // Read from Info.plist or environment
        Bundle.main.object(forInfoDictionaryKey: "SUPABASE_URL") as? String ?? "https://your-project.supabase.co"
    }
}

File: shifai-android/app/src/main/java/com/shifai/data/sync/SyncEngine.kt
package com.shifai.data.sync

import android.content.Context
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.withContext
import java.io.ByteArrayOutputStream
import java.net.HttpURLConnection
import java.net.URL
import java.security.KeyStore
import java.security.MessageDigest
import java.util.*
import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec

/**
 * Sync Engine — Android (S7-2)
 * Zero-knowledge, offline-first cloud sync
 * Flow: local DB → JSON → AES-256-GCM encrypt → blob → Supabase
 */
class SyncEngine private constructor(private val context: Context) {

    companion object {
        @Volatile
        private var instance: SyncEngine? = null

        fun getInstance(context: Context): SyncEngine {
            return instance ?: synchronized(this) {
                instance ?: SyncEngine(context.applicationContext).also { instance = it }
            }
        }

        private const val KEY_ALIAS = "shifai_sync_key"
        private const val KEYSTORE = "AndroidKeyStore"
        private const val MAX_BLOB_SIZE = 10 * 1024 * 1024 // 10MB
        private const val GCM_TAG_LENGTH = 128
        private const val GCM_IV_LENGTH = 12
    }

    sealed class SyncState {
        object Idle : SyncState()
        object Syncing : SyncState()
        data class Success(val lastSync: Date) : SyncState()
        data class Error(val message: String) : SyncState()
    }

    private val _state = MutableStateFlow<SyncState>(SyncState.Idle)
    val state: StateFlow<SyncState> = _state

    private val prefs = context.getSharedPreferences("shifai_sync", Context.MODE_PRIVATE)

    var isEnabled: Boolean
        get() = prefs.getBoolean("sync_enabled", false)
        set(value) { prefs.edit().putBoolean("sync_enabled", value).apply() }

    val lastSyncDate: Date?
        get() = prefs.getLong("last_sync_time", -1L).let {
            if (it > 0) Date(it) else null
        }

    private val syncEndpoint: String
        get() {
            val url = context.packageManager
                .getApplicationInfo(context.packageName, android.content.pm.PackageManager.GET_META_DATA)
                .metaData?.getString("SUPABASE_URL") ?: "https://your-project.supabase.co"
            return "$url/functions/v1/sync-data"
        }

    private var isSyncing = false

    // MARK: - Push

    suspend fun push() = withContext(Dispatchers.IO) {
        if (!isEnabled || isSyncing) return@withContext
        isSyncing = true
        _state.value = SyncState.Syncing

        try {
            // 1. Serialize
            val payload = serializeLocalData()

            // 2. Encrypt
            val encrypted = encrypt(payload)

            // 3. Checksum
            val checksum = sha256(encrypted)

            if (encrypted.size > MAX_BLOB_SIZE) {
                _state.value = SyncState.Error("Données trop volumineuses")
                return@withContext
            }

            // 4. Push
            val currentVersion = prefs.getInt("blob_version", 0)
            val conn = URL(syncEndpoint).openConnection() as HttpURLConnection
            conn.requestMethod = "POST"
            conn.setRequestProperty("Content-Type", "application/octet-stream")
            conn.setRequestProperty("Authorization", "Bearer ${getAuthToken()}")
            conn.setRequestProperty("X-Checksum-SHA256", checksum)
            conn.setRequestProperty("X-Blob-Version", "${currentVersion + 1}")
            conn.doOutput = true
            conn.outputStream.use { it.write(encrypted) }

            if (conn.responseCode == 200) {
                prefs.edit()
                    .putInt("blob_version", currentVersion + 1)
                    .putLong("last_sync_time", System.currentTimeMillis())
                    .apply()
                _state.value = SyncState.Success(Date())
            } else {
                _state.value = SyncState.Error("Push failed: ${conn.responseCode}")
            }
        } catch (e: Exception) {
            _state.value = SyncState.Error(e.message ?: "Unknown error")
        } finally {
            isSyncing = false
        }
    }

    // MARK: - Pull

    suspend fun pull() = withContext(Dispatchers.IO) {
        if (!isEnabled || isSyncing) return@withContext
        isSyncing = true
        _state.value = SyncState.Syncing

        try {
            // 1. Check metadata
            val metaConn = URL("$syncEndpoint?action=metadata").openConnection() as HttpURLConnection
            metaConn.setRequestProperty("Authorization", "Bearer ${getAuthToken()}")
            if (metaConn.responseCode != 200) {
                _state.value = SyncState.Error("Metadata check failed")
                return@withContext
            }

            val metaJson = metaConn.inputStream.bufferedReader().readText()
            // Parse version from JSON (simplified)
            val serverVersion = Regex("\"blob_version\":(\\d+)").find(metaJson)?.groupValues?.get(1)?.toIntOrNull() ?: 0
            val localVersion = prefs.getInt("blob_version", 0)

            if (serverVersion <= localVersion) {
                _state.value = SyncState.Success(lastSyncDate ?: Date())
                return@withContext
            }

            // 2. Pull blob
            val blobConn = URL("$syncEndpoint?action=pull").openConnection() as HttpURLConnection
            blobConn.setRequestProperty("Authorization", "Bearer ${getAuthToken()}")
            val blobData = blobConn.inputStream.use { input ->
                ByteArrayOutputStream().use { output ->
                    input.copyTo(output)
                    output.toByteArray()
                }
            }

            // 3. Verify checksum
            val serverChecksum = blobConn.getHeaderField("X-Checksum-SHA256") ?: ""
            val localChecksum = sha256(blobData)
            if (serverChecksum.isNotEmpty() && serverChecksum != localChecksum) {
                _state.value = SyncState.Error("Integrity check failed")
                return@withContext
            }

            // 4. Decrypt
            val decrypted = decrypt(blobData)

            // 5. Merge
            mergeWithLocal(decrypted)

            prefs.edit()
                .putInt("blob_version", serverVersion)
                .putLong("last_sync_time", System.currentTimeMillis())
                .apply()
            _state.value = SyncState.Success(Date())

        } catch (e: Exception) {
            _state.value = SyncState.Error(e.message ?: "Unknown error")
        } finally {
            isSyncing = false
        }
    }

    suspend fun sync() {
        push()
        pull()
    }

    // MARK: - Serialization

    private fun serializeLocalData(): ByteArray {
        // Serialize from Room DB to JSON
        // TODO: Wire to actual Room repositories
        val json = """{"cycles":[],"symptoms":[],"insights":[],"predictions":[],"timestamp":"${Date()}"}"""
        return json.toByteArray(Charsets.UTF_8)
    }

    private fun mergeWithLocal(data: ByteArray) {
        // Last-write-wins merge
        // TODO: Implement full merge with conflict detection
    }

    // MARK: - Encryption (AES-256-GCM via Android Keystore)

    private fun encrypt(data: ByteArray): ByteArray {
        val key = getOrCreateKey()
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, key)
        val iv = cipher.iv
        val encrypted = cipher.doFinal(data)
        // Prepend IV to ciphertext
        return iv + encrypted
    }

    private fun decrypt(data: ByteArray): ByteArray {
        val key = getOrCreateKey()
        val iv = data.sliceArray(0 until GCM_IV_LENGTH)
        val ciphertext = data.sliceArray(GCM_IV_LENGTH until data.size)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, key, GCMParameterSpec(GCM_TAG_LENGTH, iv))
        return cipher.doFinal(ciphertext)
    }

    private fun getOrCreateKey(): SecretKey {
        val keyStore = KeyStore.getInstance(KEYSTORE).apply { load(null) }
        if (keyStore.containsAlias(KEY_ALIAS)) {
            return (keyStore.getEntry(KEY_ALIAS, null) as KeyStore.SecretKeyEntry).secretKey
        }

        val keyGen = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, KEYSTORE)
        keyGen.init(
            KeyGenParameterSpec.Builder(KEY_ALIAS, KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT)
                .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
                .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
                .setKeySize(256)
                .build()
        )
        return keyGen.generateKey()
    }

    // MARK: - Helpers

    private fun sha256(data: ByteArray): String {
        return MessageDigest.getInstance("SHA-256")
            .digest(data)
            .joinToString("") { "%02x".format(it) }
    }

    private fun getAuthToken(): String {
        return prefs.getString("supabase_access_token", "") ?: ""
    }
}
